multinet/migrations/__init__.py

```diff
@@ -0,0 +1,54 @@
+"""Logic to manage and run migrations."""
+import pkgutil
+from functools import lru_cache
+
+from multinet.migrations.types import Migration
+from multinet.migrations.util import (
+    get_migrations,
+    get_stored_migrations,
+    store_migration,
+)
+
+from types import ModuleType
+from typing import List
+
+
+# This function must exist in this file, as it needs access to the `__path__` variable,
+# and will infinitely recurse (through pkgutil.walk_packages) if it's placed in a
+# sub-module of this module.
+@lru_cache()
+def get_modules() -> List[ModuleType]:
+    """Return all migration modules."""
+    modules = [
+        loader.find_module(name).load_module(name)
+        for loader, name, _ in pkgutil.walk_packages(__path__)  # type: ignore
+    ]
+    return modules
+
+
+def get_unapplied_migrations() -> List[Migration]:
+    """Return any migrations which have yet to be applied."""
+    modules = get_modules()
+    all_migrations = get_migrations(modules)
+    applied_migrations = {migration.name for migration in get_stored_migrations()}
+
+    return [
+        migration
+        for migration in all_migrations
+        if migration.__name__ not in applied_migrations
+    ]
+
+
+def run_migrations() -> None:
+    """Run all defined migrations."""
+
+    migrations = get_unapplied_migrations()
+    if not len(migrations):
+        print("All migrations up to date!")
+        return
+
+    for migration in migrations:
+        migration.run()
```

Comment:
If the migration fails, what happens? You wouldn't want to then store the migration as having been completed, etc. I would like to make the failure case explicit here; let me know what you think.

Reply:
I agree, this is something I just hadn't addressed. I think my reasoning in this simple implementation is that if something were to go wrong, it would throw an error and halt execution. That could obviously be handled more elegantly here, so I can handle that case. I also think it'd important to return a non-zero exit code in the actual `migrate` script, if any do fail, to prevent further actions.

Reply:
9c3b77c